File: 백준/Silver/17503. 맥주 축제/맥주 축제.cc
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<vector<int>> vi;
typedef pair<int, int> pii;
typedef pair<ll, ll> pll;
typedef vector<vector<ll>> vll;
int dx[4] = { -1, 1, 0, 0 };
int dy[4] = { 0, 0, -1, 1 };

int main(void)
{
    ios_base::sync_with_stdio(0);
    cin.tie(0); cout.tie(0);
    ll n, m, k; cin >> n >> m >> k;
    vector<pll> v;
    for (int i = 0; i < k; i++)
    {
        ll x, y; cin >> x >> y;
        v.push_back({ x, y });
    }

    sort(v.rbegin(), v.rend());

    ll sum = 0;
    for (int i = 0; i < min(n, k); i++) // 선호도 제한 없음
        sum += v[i].first;

    if (sum < m)
    {
        cout << -1;
        return 0;
    }

    ll l = 0, r = 1e15;

    while (l + 1 < r) // (l, r]
    {
        ll mid = (l + r) / 2;
        ll res = 0, cnt = 0;
        for (int i = 0; i < k; i++)
        {
            if (v[i].second <= mid)
            {
                res += v[i].first;
                cnt++;
            }

            if (cnt == n)
                break;
        }

        if (res >= m && cnt == n) r = mid;
        else l = mid;
    }

    cout << r;
}

File: 백준/Silver/17503. 맥주 축제/맥주 축제.py
# 입력 받기
n, m, k = map(int, input().split())

# (x, y) 쌍을 입력받아 리스트에 저장
pairs = []
for _ in range(k):
    x, y = map(int, input().split())
    pairs.append((x, y))

# (x, y) 쌍을 x 기준 내림차순으로 정렬
pairs.sort(reverse=True, key=lambda pair: pair[0])

# 상위 n개의 x 값의 합이 m을 넘는지 확인
if sum(x for x, y in pairs[:n]) < m:
    print(-1)
else:
    # 이진 탐색으로 조건을 만족하는 최소 y 찾기
    l, r = 0, 10**15
    while l + 1 < r:
        mid = (l + r) // 2
        res, cnt = 0, 0

        for x, y in pairs:
            if y <= mid:
                res += x
                cnt += 1
                if cnt == n:
                    break

        if cnt == n and res >= m:
            r = mid
        else:
            l = mid

    # 가능한 최소 y 값 출력
    print(r)
